File: C/AR_104.c
#include <stdio.h>

int movil(int* sumNodos, int* fin) {
    int pi=0, di=0, pd=0, dd=0;

    scanf("%d %d %d %d", &pi, &di, &pd, &dd);
    if (pi==0 && di==0 && pd==0 && dd==0) {
        *fin=1;
        return 0;
    }

    // CASO BASE
    if (pi!=0 && pd!=0) { // MOVIL SIN SUBMOVILES
        *sumNodos+=pi+pd;
        if (pi*di==pd*dd) return 1;
        else return 0;
    }

    // CASO RECURSIVO    
    // IZQ
    int izq=1;
    int izqNodos=pi;
    if (pi==0) {
        izq=movil(&izqNodos, fin);
    }

    // DERECHA
    int der=1;
    int derNodos=pd;
    if (pd==0) der=movil(&derNodos, fin);
    *sumNodos=izqNodos+derNodos;

    return izqNodos*di==derNodos*dd && izq && der;
}

int solve() {
    int sumNodos=0;
    int fin=0;

    if (movil(&sumNodos, &fin)) printf("SI\n");
    else if (fin) return 0;
    else printf("NO\n");
    
    return 1;
}

int main() {
    while (solve()) {}
    return 0;
}

File: Java/AR_104.java
import java.util.Scanner;

public class Solution {
    static Scanner scanner = new Scanner(System.in);

    static class Result {
        int sumNodos;
        boolean fin;
        boolean balanced;

        Result(int sumNodos, boolean fin, boolean balanced) {
            this.sumNodos =sumNodos;
            this.fin 	  =fin;
            this.balanced =balanced;
        }
    }

    static Result movil() {
        int pi =scanner.nextInt();
        int di =scanner.nextInt();
        int pd =scanner.nextInt();
        int dd =scanner.nextInt();
        if (pi==0 && di==0 && pd==0 && dd==0) {
            return new Result(0, true, false);
        }

        // CASO BASE
        if (pi!=0 && pd!=0) { // MOVIL SIN SUBMOVILES
            return new Result(pi+pd, false, pi*di==pd*dd);
        }

        // CASO RECURSIVO    
        // IZQ
        Result izq= pi==0? movil():new Result(pi, false, true);

        // DERECHA
        Result der= pd==0? movil():new Result(pd, false, true);

        int sumNodos    =izq.sumNodos+der.sumNodos;
        boolean balanced=(izq.sumNodos*di==der.sumNodos*dd) && izq.balanced && der.balanced;

        return new Result(sumNodos, izq.fin || der.fin, balanced);
    }

    static boolean solve() {
        Result result=movil();

        if (result.fin) return false;
        System.out.println(result.balanced ? "SI":"NO");
        
        return true;
    }

    public static void main(String[] args) {
        while (solve());
    }
}